pkg/kv/kvserver/spanset/merge.go

```diff
@@ -5,114 +5,57 @@
 
 package spanset
 
-import "sort"
+import "slices"
 
-type sortedSpans []Span
-
-func (s *sortedSpans) Less(i, j int) bool {
-	// Sort first on the start key and second on the end key. Note that we're
-	// relying on EndKey = nil (and len(EndKey) == 0) sorting before other
-	// EndKeys.
-	c := (*s)[i].Key.Compare((*s)[j].Key)
-	if c != 0 {
-		return c < 0
-	}
-	return (*s)[i].EndKey.Compare((*s)[j].EndKey) < 0
-}
-
-func (s *sortedSpans) Swap(i, j int) {
-	(*s)[i], (*s)[j] = (*s)[j], (*s)[i]
-}
-
-func (s *sortedSpans) Len() int {
-	return len(*s)
-}
-
-// mergeSpans sorts the given spans and merges ones with overlapping
-// spans and equal access timestamps. The implementation is a copy of
-// roachpb.MergeSpans.
+// mergeSpans sorts the given spans and merges ones with overlapping spans and
+// equal access timestamps. The implementation is a modified roachpb.MergeSpans.
 //
-// Returns true iff all of the spans are distinct.
 // The input spans are not safe for re-use.
-func mergeSpans(latches *[]Span) ([]Span, bool) {
-	if len(*latches) == 0 {
-		return *latches, true
+//
+// TODO(pav-kv): explain and assert on what this function expects. It definitely
+// does not solve the most general form of merging, but it seems the expectation
+// here is that every key is covered only by one timestamp.
+func mergeSpans(latches []Span) []Span {
+	if len(latches) == 0 {
+		return latches
 	}
-
-	sort.Sort((*sortedSpans)(latches))
+	// Sort first on the start key and second on the end key. Note that we're
+	// relying on empty EndKey sorting before other EndKeys.
+	slices.SortFunc(latches, func(s1, s2 Span) int {
+		if c := s1.Key.Compare(s2.Key); c != 0 {
+			return c
+		}
+		return s1.EndKey.Compare(s2.EndKey)
+	})
 
 	// We build up the resulting slice of merged spans in place. This is safe
 	// because "r" grows by at most 1 element on each iteration, staying abreast
 	// or behind the iteration over "latches".
-	r := (*latches)[:1]
-	distinct := true
+	r := latches[:1]
 
-	for _, cur := range (*latches)[1:] {
+	for _, cur := range latches[1:] {
 		prev := &r[len(r)-1]
-		if len(cur.EndKey) == 0 && len(prev.EndKey) == 0 {
-			if cur.Key.Compare(prev.Key) != 0 {
-				// [a, nil] merge [b, nil]
-				r = append(r, cur)
-			} else {
-				// [a, nil] merge [a, nil]
-				if cur.Timestamp != prev.Timestamp {
-					r = append(r, cur)
-				}
-				distinct = false
+		// Can only merge spans at the same timestamp.
+		if cur.Timestamp != prev.Timestamp {
+			r = append(r, cur)
+		} else if len(prev.EndKey) == 0 { // prev is a point key
+			if !cur.Key.Equal(prev.Key) { // cur.Key > prev.Key
```

Comment:
So this func goes an extra mile to merge `[a,b) + [b] = [a,b]` (see below). However, it doesn't handle cases like two consecutive point keys:

```
[a] + [a.Next()] = [a,a.Next().Next())
```

I wonder if we should do it for completeness. Today:

```
[a,b) + [b] + [b.Next()] = [a,b.Next().Next())
        [b] + [b.Next()] = [b] + [b.Next()]
```

Reply:
I have a commit that fixes it, but I'll send it separately after this PR merges. As is, the PR does not change the behaviour of `mergeSpans`.